### DISCUSSION TITLE
Creating a non-CAMELS dataset

### DISCUSSION THREAD

**Discussion by darylsuen** [neuralhydrology/neuralhydrology#181] [General]
Hi, I'm new to Python and machine learning, just finished a geospatial data analytics program. I'm building workflows for using neuralhydrology for a research department at my college. I'm creating forcing and attribute datasets for basins in Nova Scotia from scratch. I'm struggling with integrating the datasets into neuralhyrdrology. I'm trying to copy the format of the CAMELS_US datasets as best I can. 

Question 1: The forcing .txt files, like 01013500_lump_nldas_forcing_leap.txt, for example, all have the first column and first three rows dedicated to numbers. It looks like the number on the third row is the basin id but what are the first two numbers? I prefer to work with .csv instead of tab separated files. It's fine to use .csv instead?

Question 2: I've written a .py script for adding a new dataset class for the datasetzoo. Do I just add the .py script to the others in the datasetzoo directory of the neuralhydrology package? It doesn't say what to do with the class .py after writing it in the 'Adding a new Dataset' tutorial. Perhaps someone could direct me to documentation for creating datasets from scratch?

Question 4: Am I going about this in the wrong way if I'm not using CAMELS datasets i.e. should I just be using PyTorch? I wanted the researchers to be able to set up the models and adjust parameters through config files so they wouldn't have to touch the .py scripts as they don't have programming experience. Also, config files are very useful for creating a library of dataset, model, and parameter settings.

Cheers!

**Comment by kratzert** [comment]
Hi,

no need to go extra steps and trying to replicate the filestructure or the CAMELS dataset. By far the easiest solution for you is to use the `GenericDataset` class that is implemented. Check out the tutorial on how to use this dataset class here https://neuralhydrology.readthedocs.io/en/latest/tutorials/add-dataset.html#Using-the-GenericDataset

**Reply by darylsuen** [reply]
Hi Frederic. Thanks for the quick response. Yes, I have looked at that tutorial but it asks that the forcings data be in netcdf files. The netcdf files from the RDRS_v2.1 are in hourly intervals and I wasn't able to figure out how to convert to daily due to not being able to change the time coordinate array from shape[0] of 24 to just 1. Consequently, I was forced to create .csv files for the basin forcing data. I have some deadlines to meet so I haven't explored the GenericDataset option after reading the netcdf guideline to see if it could still work with .csv files. Do you know if GenericDataset works with .csv?

**Comment by darylsuen** [comment]
Does anyone know if I can use the GenericDataset class to add a custom dataset that consists of .csv files instead of netcdf files?

**Reply by gauchm** [reply]
No, GenericDataset does not support csv files. But if you have the csv, converting that into a netcdf shouldn't be hard. I recommend you take a look at the [xarray library](https://docs.xarray.dev/en/stable/), which provides a simple dataset.to_netcdf() function.

**Reply by darylsuen** [reply]
I managed to create a generic dataset but the tutorial is vague about how the columns should be set up. The CAMELS US data has the forcing variables in a separate file from the streamflow variable.

**Reply by gauchm** [reply]
https://neuralhydrology.readthedocs.io/en/latest/api/neuralhydrology.datasetzoo.genericdataset.html is actually quite explicit about the required format. All timeseries for a basin, including the streamflow, should be in the same netcdf.

**Comment by darylsuen** [comment]
I have created a new dataset class and I have put the .py script in the datasetzoo folder of the neuralhydrology package.  In the config.yml, I need to know what to enter for 'dataset:'. The examples are 'camels_us', 'camels_gb', etc. These are not the names of the .py scripts of the dataset classes in datasetzoo directory. Where are these names stored so that I can add the name of my custom dataset to the list?

**Reply by darylsuen** [reply]
I went into the conftest.py file and it looks like the string that is entered for dataset: in the config file is supposed to be mapped to the target variable, which, here, is the column name 'QObs(mm/d)'. I have created a generic dataset with a netcdf4 file for a single basin containing a time series of forcing variables including the target variable, stream discharge rate. I don't know if I was supposed to name the column of my target variable with the same name, 'QObs(mm/d)', in order for it to be recognized. It doesn't specify in the tutorial how the columns should be set up. I noticed in the CAMELS data that the forcings variables are in a separate dataset form the streamflow variable, which is in a dataset by itself. Has anyone been successful in creating a generic dataset?

@pytest.fixture(params=[('camels_us', ['QObs(mm/d)'])], ids=lambda param: param[0])
def daily_dataset(request) -> Dict[str, List[str]]:
    """Fixture that provides daily datasets.

    Returns
    -------
    Dict[str, List[str]]
        Dictionary ``{'dataset: <name of the dataset>, 'target': <list of target variables>}``.
    """

**Reply by gauchm** [reply]
The various dataset names for the config are defined here: https://github.com/neuralhydrology/neuralhydrology/blob/master/neuralhydrology/datasetzoo/__init__.py#L62-L80

For the GenericDataset, you want to use `"generic"`.

The variable names are entirely up to you (although note that IIRC netcdf doesn't support certain characters). You specify in the config.yaml which variables are to be used as forcings and which ones are to be used as target variables.
